yt_dlp/extractor/glomex.py

```diff
@@ -0,0 +1,224 @@
+
+    def _extract_api_data(self, video, video_id):
+        if video.get('error_code') == 'contentGeoblocked':
+            self.raise_geo_restricted(countries=video['geo_locations'])
+
+        formats, subs = [], {}
+        for format_id, format_url in video['source'].items():
+            ext = determine_ext(format_url)
+            if ext == 'm3u8':
+                formats_, subs_ = self._extract_m3u8_formats_and_subtitles(
+                    format_url, video_id, 'mp4', m3u8_id=format_id,
+                    fatal=False)
+                formats.extend(formats_)
+                subs.update(subs_)
+            else:
+                formats.append({
+                    'url': format_url,
+                    'format_id': format_id,
+                })
+        if video.get('language'):
+            for fmt in formats:
+                fmt['language'] = video['language']
+        self._sort_formats(formats)
+
+        thumbnails = [{
+            'id': image.get('id'),
+            'url': f'{image["url"]}/profile:player-960x540',
+            'width': 960,
+            'height': 540,
+        } for image in (video.get('images') or []) + [video.get('image') or {}] if image.get('url')]
+        self._remove_duplicate_formats(thumbnails)
+
+        return {
+            'id': video.get('clip_id') or video_id,
+            'title': video.get('title'),
+            'description': video.get('description'),
+            'thumbnails': thumbnails,
+            'duration': int_or_none(video.get('clip_duration')),
+            'timestamp': video.get('created_at'),
+            'formats': formats,
+            'subtitles': subs,
+        }
+
+
+class GlomexIE(GlomexBaseIE):
+    IE_NAME = 'glomex'
+    IE_DESC = 'Glomex videos'
+    _VALID_URL = r'https?://video\.glomex\.com/[^/]+/(?P<id>v-[^-]+)'
+    _INTEGRATION_ID = '19syy24xjn1oqlpc'
+
+    _TESTS = [{
+        'url': 'https://video.glomex.com/sport/v-cb24uwg77hgh-nach-2-0-sieg-guardiola-mit-mancity-vor-naechstem-titel',
+        'md5': 'cec33a943c4240c9cb33abea8c26242e',
+        'info_dict': {
+            'id': 'v-cb24uwg77hgh',
+            'ext': 'mp4',
+            'title': 'md5:38a90cedcfadd72982c81acf13556e0c',
+            'description': 'md5:1ea6b6caff1443fcbbba159e432eedb8',
+            'duration': 29600,
+            'timestamp': 1619895017,
+            'upload_date': '20210501',
+            'age_limit': None,
+        },
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        return self.url_result(
+            GlomexEmbedIE.build_player_url(video_id, self._INTEGRATION_ID, url),
+            GlomexEmbedIE.ie_key(), video_id)
+
+
+class GlomexEmbedIE(GlomexBaseIE):
+    IE_NAME = 'glomex:embed'
+    IE_DESC = 'Glomex embedded videos'
+    _BASE_PLAYER_URL = '//player.glomex.com/integration/1/iframe-player.html'
+    _VALID_URL = rf'https?:{re.escape(_BASE_PLAYER_URL)}\?([^#]+&)?playlistId=(?P<id>[^#&]+)'
```

Comment:
> _VALID_URL would match anything in the integration version field, rather than 1, which is hard-coded in _BASE_PLAYER_URL.

```py
        # make the scheme in _VALID_URL optional
        _URL_RE = r'(?:https?:)?//' + cls._VALID_URL.split('://', 1)[1]
        # simplify the query string part of _VALID_URL; after extracting iframe
        # src, the URL will be matched again
        _URL_RE = _URL_RE.split(r'\?', 1)[0] + r'\?(?:(?!(?P=_q1)).)+'
```

I wanted to avoid this mess in _extract_urls and also not have to duplicate the base URL everywhere, and totally missed the generalization of version number. Is it just for generalization? or is there actual URLs that match it? While I all am for generalizing valid_url for future-proofing, if it has no actual use, I dont think the added complexity is worth it. If it is actually needed, we can revert

---

> It is also does not capture integration parameter. Capturing a second parameter, allowing other parameters and arbitrary order considerably complicates things, so I understand you opted to do this later during IE extraction. It would be good to at least add a comment here to that effect.

imo, the error is sufficient self-documentation
```
        integration = parse_qs(url).get('integrationId', [None])[0]
        if not integration:
            raise ExtractorError('No integrationId in URL', expected=True)
```

Reply:
> I wanted to avoid this mess in _extract_urls and also not have to duplicate the base URL everywhere, and totally missed the generalization of version number. Is it just for generalization? or is there actual URLs that match it? While I all am for generalizing valid_url for future-proofing, if it has no actual use, I dont think the added complexity is worth it. If it is actually needed, we can revert

In practice if you request /integration/1 you get a redirect to /integration/1.xyz (some current minor version). However there is nothing to stop any site from embedding using this final exact version. I believe it is necessary, so I will try to add it back.

> imo, the error is sufficient self-documentation

But the error is way down below in the last method -- the comment is right there to explain things to the reader. I would like to add it back if you don't mind.

Reply:
sure